controller/lib/core/blower_fsm.cpp

```diff
@@ -105,34 +137,40 @@ BlowerSystemState PressureAssistFsm::DesiredState() const {
 }
 
 // TODO don't rely on fsm inner states to make this usable in any fsm
-bool PressureAssistFsm::PatientInspiring(const BlowerFsmInputs &inputs) {
-  // TODO: change this when we get better understanding of inspiratory effort
-  // detection strategy, right now this is just set to trigger for any flow
-  // that is higher than twice the flowmeter's flow noise :
-  //   flow = flow_inhale (+/- noise) - flow_exhale (+/- noise).
-  // According to data from Ingmar test lung
-  // (https://drive.google.com/open?id=11_A8KEDdBa5l7rqUegfKsj71rug1l7Sx),
-  // this should work OK as a first approximation.
-  if (inspire_finished_ && inputs.net_flow >= ml_per_min(0) &&
-      inspiratory_effort_threshold_ >= cubic_m_per_sec(9)) {
-    // Not sure how to go about this: when we enter this breath fsm, either we
-    // were Off and flow should be 0, which means this is OK: the first
-    // positive flow we see is either noise or the blower starting to provide
-    // PEEP we can take that as an  "almost zero" reference flow, or we are at
-    // PIP, I assume the flow will be negative until we reach PEEP, where it
-    // will approximately be 0 and we may assume to be back to the first case.
-    inspiratory_effort_threshold_ = inputs.net_flow + ml_per_min(12000.0f);
-    // Note 12000 ml/min looks like a lot but it is actually 200 ml/sec which
-    // is in the high end of our sensor's sensitivity for relatively low flows
-    // that are expected during exhale, especially when summing two sensors.
-    // I can see two ways to improve this sensitivity:
-    //   a: provide a denoised flow in SensorReadings, at the cost of reducing
-    //      response time
-    //   b: provide separate inflow and outflow, at the cost of a more
-    //      complicated (and not yet determined) breath detection algorithm.
+bool PressureAssistFsm::PatientInspiring(Time now,
+                                         const BlowerFsmInputs &inputs) {
+  if (!inspire_finished_ || inputs.net_flow < ml_per_sec(0)) {
+    return false;
   }
 
-  return inspire_finished_ && inputs.net_flow > inspiratory_effort_threshold_;
+  // Once we're done inspiring and flow is non-negative, start calculating two
+  // exponentially-weighted averages of net flow: slow_avg_flow_ and
+  // fast_avg_flow_.
+  //
+  // The slow one has a smaller alpha term, so updates slower than the fast
+  // one.  You can think of the slow average as estimating "flow at dwell" and
+  // the fast average as estimating "current flow".
+  //
+  // If the fast average exceeds the slow average, we trigger a breath.
+  float slow_alpha = dbg_pa_slow_avg_alpha.Get();
+  float fast_alpha = dbg_pa_fast_avg_alpha.Get();
+
+  // TODO: This could be encapsulated in an exponentially-weighted-average
+  // class.
+  if (slow_avg_flow_ == std::nullopt) {
```

Comment:
You can replace these with .value_or() at lines 167 and 169.

Reply:
Much better, thanks!